crates/esplora/src/blocking_ext.rs

```diff
@@ -68,250 +56,368 @@ pub trait EsploraExt {
-                            Some((
-                                OutPoint {
-                                    txid: vin.txid,
-                                    vout: vin.vout,
-                                },
-                                TxOut {
-                                    script_pubkey: prevout.scriptpubkey.clone(),
-                                    value: prevout.value,
-                                },
-                            ))
-                        });
-
-                        for (outpoint, txout) in previous_outputs {
-                            let _ = graph.insert_txout(outpoint, txout);
-                        }
-                    }
-                }
+        let fetched_hash = match fetched_blocks.entry(height) {
+            btree_map::Entry::Occupied(entry) => *entry.get(),
+            btree_map::Entry::Vacant(entry) => *entry.insert(client.get_block_hash(height)?),
+        };
 
-                let last_index = last_index.expect("Must be set since handles wasn't empty.");
-                let past_gap_limit = if let Some(i) = last_active_index {
-                    last_index > i.saturating_add(stop_gap as u32)
-                } else {
-                    last_index >= stop_gap as u32
-                };
-                if past_gap_limit {
-                    break;
-                }
+        // We have found point of agreement so the update will connect!
+        if fetched_hash == local_hash {
+            break;
+        }
+    }
+
+    Ok(fetched_blocks)
+}
+
+/// Fetches missing checkpoints and finalizes the [`local_chain::Update`].
+///
+/// A checkpoint is considered "missing" if an anchor (of `anchors`) points to a height without an
+/// existing checkpoint/block under `local_tip` or `update_blocks`.
+#[doc(hidden)]
+pub fn finalize_chain_update_blocking<A: Anchor>(
+    client: &esplora_client::BlockingClient,
+    local_tip: &CheckPoint,
+    anchors: &BTreeSet<(A, Txid)>,
+    mut update_blocks: BTreeMap<u32, BlockHash>,
+) -> Result<local_chain::Update, Error> {
+    let update_tip_height = update_blocks
+        .keys()
+        .last()
+        .copied()
+        .expect("must atleast have one block");
+
+    // We want to have a corresponding checkpoint per height. We iterate the heights of anchors
+    // backwards, comparing it against our `local_tip`'s chain and our current set of
+    // `update_blocks` to see if a corresponding checkpoint already exists.
+    let anchor_heights = anchors
+        .iter()
+        .map(|(a, _)| a.anchor_block().height)
+        // filter out duplicate heights
+        .filter({
+            let mut prev_height = Option::<u32>::None;
+            move |h| match prev_height.replace(*h) {
+                None => true,
+                Some(prev_h) => prev_h != *h,
             }
+        })
+        // filter out heights that surpass the update tip
+        .filter(|h| *h <= update_tip_height)
+        .rev();
```

Comment:
:pick: I feel it's better to put the `.rev` first unless there's some reason it has to be last. The comment above says we're iterating backwards but it seems to by lying.

:pick: Can the `.filter` be `.take_while` here?

Reply:
The `.filter` cannot be `.take_while` here since we are only filtering out top-most (highest) blocks (if there is anything to filter out at all).